### DISCUSSION TITLE
Trying to process a file when opening and when saving

### DISCUSSION THREAD

**Discussion by rijuma** [microsoft/vscode-discussions#1236] [Extension Development QnA] [answered]
Hello. I want to do an extension to make it easy for me to work with some encoded text files.

The idea is that, when loading in VSCode, the contents of the file gets decoded, so the editor sees a text file, and whenever the file is saved, the process is reversed and the resulting file is encoded.

One way to do it would be to use `vscode.workspace.onDidOpenTextDocument()` to check the document extension and if it matches my extension then use `vscode.WorkspaceEdit()` to update its contents, but this would trigger an update on the file, so it will show as "dirty" just by opening it, which is not ideal.

The saving part I think it's easier since I could use `vscode.workspace.onWillSaveTextDocument()` to make my transformations to the contents before saving.

To explain the behavior in a simple way: For the user, editing these files should behave exactly like editing a normal text file, the encoding/decoding should be invisible to the user.

I've been reading about [TextDocumentContentProvider](https://code.visualstudio.com/api/extension-guides/virtual-documents#textdocumentcontentprovider), which sounds like a possibility, but I'm not sure on how to keep the normal behavior on opening the files with the file explorer. Also there's this [FileSystemProvider](https://code.visualstudio.com/api/references/vscode-api#FileSystemProvider) as well, but both solutions sound way too convoluted for this simple task.

Any examples are much welcome.

**Comment by rijuma** [comment]
Reading a bit further into this I've found some similar functionality of what I want to do on the NotebookSerializer, since you can use a custom code to seralize and deserialize the notebook contents, but I don't want all the Notebook block functionality.

I would basically need to set up a serializer for my custom file types, so it reads as a text file but it saves as a hashed serialized text.

Still searching for a way to do this.

**Reply by gjsjohnmurray** [reply, marked as answer]
I recommend the FileSystemProvider route.

**Reply by rijuma** [reply]
> I recommend the FileSystemProvider route.

Thanks for your suggestion. I tried, but the examples were too complicated (for me at least), since it creates a whole virtual file system. Do you have some easier example when I could use the existing file system but just add some pipeline in between the opening and saving of the current file?

I've also looked at the [VSCode HexEditor extension](https://github.com/microsoft/vscode-hexeditor) but you need to explicitly open the file with the editor. in this case the idea would be just to do a seamlessly integration.
